### PR TITLE
perf: Reworked types and rendering to try and reduce needless renders

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

Notebooks re-render **a lot**. Turns out this was due to a silly bit of rendering logic that was hard to track down.

This PR adds some tooling _and_ fixes some of the perf issues

## Changes

* Adds a `useWhyDidIRender` hook which is flagged, allowing us to turn on and off logging as to which props are causing re-renders. 
* Adds a `useFeatureFlag` helper (I should have done this in another PR but I finally got bored of writing the same 3 lines of code) 
* Used this to find out that the `children` prop was the offender and if we passed the Component in as a prop instead and then only passed the two properties they need - everything performs way better!

Not sure if this is clear but via React dev tools detecting re-renders, before you can see every type causes the underlying Nodes to render their children. Now it only renders the parent. Probably still improvements to be done but this is progress!

|Before|After|
|----|----|
|  |  |

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Review by daibhin — APPROVED** [review]
Fantastic find! Type changes seem pretty reasonable and everything seems to work as before 🚢

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx [hunk 1]
```diff
@@ -99,7 +111,7 @@ export function Notebook({ shortId, editable = false, initialAutofocus = null }:
                     <NotebookSidebar />
                     <ErrorBoundary>
                         <Editor
-                            initialContent={content}
+                            initialContent={{}}
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:114]: This is one of the needed changes. With every change to content the whole component rerenders. Instead doing it in the logic

[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:114]: I'll add a comment to make that clearer

File: frontend/src/scenes/notebooks/Nodes/NodeWrapper.tsx [hunk 2]
```diff
@@ -279,17 +297,13 @@ export function createPostHogWidgetNode<T extends CustomNotebookNodeAttributes>(
...
-        const nodeProps: NotebookNodeViewProps<T> = {
+        const nodeProps: NotebookNodeProps<T> & Omit<NodeViewProps, 'attributes' | 'updateAttributes'> = {
             ...props,
             attributes,
             updateAttributes,
         }
 
-        return (
-            <NodeWrapper {...nodeProps} {...wrapperProps}>
-                <Component {...nodeProps} />
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Nodes/NodeWrapper.tsx:290]: Not super sure. It's more that we can't memoize the node due to the children being "dynamic" in this case

[the remark this answers is not in the document]
Comment (daibhin) [reply frontend/src/scenes/notebooks/Nodes/NodeWrapper.tsx:290]: Ah interesting, fair point

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (daibhin) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:114]: Ah ok, can we remove the prop entirely from `Editor` in that case?

[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:114]: Good point

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx [hunk 3]
```diff
@@ -58,16 +70,16 @@ export function Notebook({ shortId, editable = false, initialAutofocus = null }:
...
-    } else if (!notebook) {
+    } else if (!notebookLoaded) {
         return <NotFound object="notebook" />
     }
 
     return (
         <BindLogic logic={notebookLogic} props={{ shortId }}>
             <div className={clsx('Notebook', !isExpanded && 'Notebook--compact', editable && 'Notebook--editable')}>
-                {notebook.is_template && (
+                {false && (
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:82]: Everytime we save, we update the notebook via an api request which means the "notebook" object changed. Kea magically selects the value and triggers a re-render if the equality changes, meaning everytime we save the `useValues` return will change.

[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:82]: So what you're proposing wouldn't work, no. We would need a selector just for the `is_template` which is what I'm going to add

**Comment by benjackwhite** [comment]
@daibhin I actually put back in the normal "notebook" value. Turns out there were no real performance gains from abstracting out. All of the gains came from the optimisation of the custom nodes

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx [hunk 4]
```diff
@@ -34,8 +35,20 @@ export function Notebook({ shortId, editable = false, initialAutofocus = null }:
...
+        isEditable,
+        shortId,
+        editable,
+        initialAutofocus,
+        content,
+    })
+
     useEffect(() => {
-        if (!notebook && !notebookLoading) {
+        if (notebook && !notebookLoading) {
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Notebook/Notebook.tsx:51]:
```suggestion
        if (!notebook && !notebookLoading) {
```

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx [hunk 5]
```diff
@@ -25,7 +26,7 @@ export type NotebookProps = {
 
 const PLACEHOLDER_TITLES = ['Release notes', 'Product roadmap', 'Meeting notes', 'Bug analysis']
 
-export function Notebook({ shortId, editable = false, initialAutofocus = null }: NotebookProps): JSX.Element {
+export function Notebook({ shortId, editable = false, initialAutofocus = 'start' }: NotebookProps): JSX.Element {
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Notebook/Notebook.tsx:29]: Is there a reason to change this value? I don't remember the specific example but I think I tried setting a default before and it caused unexpected behaviour in some places where the popover would open and you couldn't just start typing to append to the end of a notebook

File: frontend/src/scenes/notebooks/Notebook/Notebook.tsx (hunk 4 above)
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Notebook.tsx:51]: Oops merged a bit early. I'll follow up with a fix

**Pull request merged**